python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -156,1253 +120,308 @@ class MakeBrighterFatterKernelTaskConfig(pexConfig.Config):
+                Resulting Brighter-Fatter Kernel.
+        """
+        if len(dummy) == 0:
+            self.log.warn("No dummy exposure found.")
 
-        dataRef.put(bfKernel)
+        detector = camera[inputDims['detector']]
+        detName = detector.getName()
 
-        self.log.info('Finished generating kernel(s) for %s' % detNum)
-        return pipeBase.Struct(exitStatus=0)
+        if self.config.level == 'DETECTOR':
+            detectorCorrList = list()
+
+        bfk = BrighterFatterKernel(camera=camera, detectorId=detector.getId(), level=self.config.level)
+        bfk.means = inputPtc.finalMeans  # ADU
+        bfk.variances = inputPtc.finalVars  # ADU^2
+        bfk.rawXcorrs = inputPtc.covariances  # ADU^2
+        bfk.badAmps = inputPtc.badAmps
+        bfk.shape = (inputPtc.covMatrixSide*2 + 1, inputPtc.covMatrixSide*2 + 1)
+        bfk.gain = inputPtc.gain
+        bfk.noise = inputPtc.noise
+        bfk.meanXcorrs = dict()
+        bfk.valid = dict()
 
-    def _applyGains(self, means, xcorrs, ptcData):
-        """Apply the gains calculated by the PtcTask.
+        for amp in detector:
+            ampName = amp.getName()
+            mask = np.array(inputPtc.expIdMask[ampName], dtype=bool)
+
+            gain = bfk.gain[ampName]
+            fluxes = np.array(bfk.means[ampName])[mask]
+            variances = np.array(bfk.variances[ampName])[mask]
+            xCorrList = [np.array(xcorr) for xcorr in bfk.rawXcorrs[ampName]]
+            xCorrList = np.array(xCorrList)[mask]
+
+            if gain <= 0:
+                # We've received very bad data.
+                self.log.warn("Impossible gain recieved from PTC for %s: %f.  Skipping amplifier.",
+                              ampName, gain)
+                bfk.meanXcorrs[ampName] = np.zeros(bfk.shape)
+                bfk.ampKernels[ampName] = np.zeros(bfk.shape)
+                bfk.valid[ampName] = False
+                continue
 
-        It also removes datapoints that were thrown out in the PTC algorithm.
+            fluxes = np.array([flux*gain for flux in fluxes])  # Now in e^-
+            variances = np.array([variance*gain*gain for variance in variances])  # Now in e^2-
+
+            # This should duplicate the else block in generateKernel@L1358,
+            # which in turn is based on Coulton et al Equation 22.
+            scaledCorrList = list()
+            for xcorrNum, (xcorr, flux, var) in enumerate(zip(xCorrList, fluxes, variances), 1):
+                q = np.array(xcorr) * gain * gain  # xcorr now in e^-
+                q *= 2.0  # Remove factor of 1/2 applied in PTC.
+                self.log.info("Amp: %s %d/%d Flux: %f  Var: %f  Q(0,0): %g  Q(1,0): %g  Q(0,1): %g",
+                              ampName, xcorrNum, len(xCorrList), flux, var, q[0][0], q[1][0], q[0][1])
+
+                # Normalize by the flux, which removes the (0,0)
+                # component attributable to Poisson noise.
+                q[0][0] -= 2.0*(flux)
```

Comment:
What's the origin of the factor of two here? Is it because of the pair of flats at the same exposure time? Is this subtracting the first two terms in Equation 24 of Coulton+17?

Reply:
This is nominally the sum of the flux levels of the two flats.  From `cpExtractPtcTask.py` L373, the rawMeans is set to `0.5*(mu1 + mu2)` so this factor of two removes the 0.5 there.